File: src/bwlabel.cc
//Copyright (C) 2002 Jeffrey E. Boyd <boyd@cpsc.ucalgary.ca>
//
// This program is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program; if not, see <http://www.gnu.org/licenses/>.

#include <oct.h>
#include <algorithm>

#define     NO_OBJECT       0

static int find (int set [], int x)
{
  int r = x;
  while (set [r] != r)
    r = set [r];
  return r;
}

static bool any_bad_argument (const octave_value_list& args)
{
  const int nargin = args.length ();
  if (nargin < 1 || nargin > 2)
    {
      print_usage ();
      return true;
    }

  if (nargin == 2)
    {
      if (!args (1).is_real_scalar ())
        {
          error ("bwlabel: second input argument must be a real scalar");
            return true;
        }
      const int n = args (1).int_value ();
      if (n != 4 && n != 6 && n != 8)
        {
          error ("bwlabel: second input argument bust be either 4, 6 or 8");
          return true;
        }
    }

  return false;
}

DEFUN_DLD(bwlabel, args, , "\
-*- texinfo -*-\n\
@deftypefn {Loadable Function} {[@var{l}, @var{num}] =} bwlabel(@var{bw}, @var{n})\n\
Labels foreground objects in the binary image @var{bw}.\n\
The output @var{l} is a matrix where 0 indicates a background pixel,\n\
1 indicates that the pixel belong to object number 1, 2 that the pixel\n\
belong to object number 2, etc.\n\
The total number of objects is @var{num}.\n\
\n\
To pixels belong to the same object if the are neighbors. By default\n\
the algorithm uses 8-connectivity to define a neighborhood, but this\n\
can be changed through the argument @var{n} that can be either 4, 6, or 8.\n\
\n\
The algorithm is derived from  BKP Horn, Robot Vision, MIT Press,\n\
1986, p 65 - 89\n\
@end deftypefn\n\
")
{
  octave_value_list rval;
  if (any_bad_argument (args))
    return rval;

  // input arguments
  boolMatrix BW = args (0).bool_array_value ();     // the input binary image
  int nr = BW.rows ();
  int nc = BW.columns ();

  // n-hood connectivity
  int n;
  if (args.length () < 2)
    n = 8;
  else
    n = args (1).int_value ();

  // results
  Matrix L (nr, nc);     // the label image
  int nobj;              // number of objects found in image

  // other variables
  int ntable;            // number of elements in the component table/tree

  OCTAVE_LOCAL_BUFFER (int, lset, nr * nc);   // label table/tree
  ntable = 0;
  lset [0] = 0;

  for (int r = 0; r < nr; r++)
    {
      for (int c = 0; c < nc; c++)
        {
          if (BW.elem (r, c)) // if A is an object
            {
              // get the neighboring pixels B, C, D, and E
              int B, C, D, E;
              if (c == 0)
                B = 0;
              else
                B = find (lset, (int)L.elem (r, c-1));

              if (r == 0)
                C = 0;
              else
                C = find (lset, (int)L.elem (r-1, c));

              if (r == 0 || c == 0)
                D = 0;
              else
                D = find (lset, (int)L.elem (r-1, c-1));

              if (r == 0 || c == nc - 1)
                E = 0;
              else
                E = find (lset, (int)L.elem(r-1, c+1));

              if (n == 4)
                {
                  // apply 4 connectedness
                  if (B && C) // B and C are labeled
                    {
                      if ( B == C )
                        {
                          L.elem (r, c) = B;
                        }
                      else
                        {
                          lset [C] = B;
                          L.elem (r, c) = B;
                        }
                      }
                    else if (B) // B is object but C is not
                      {
                        L.elem (r, c) = B;
                      }
                    else if (C) // C is object but B is not
                      {
                        L.elem (r, c) = C;
                      }
                    else // B, C, D not object - new object
                      {
                        // label and put into table
                        ntable ++;
                        L.elem (r, c) = lset [ntable] = ntable;
                      }
                }
              else if (n == 6)
                {
                  // apply 6 connected ness
                  if (D) // D object, copy label and move on
                    {
                      L.elem (r, c) = D;
                    }
                  else if (B && C) // B and C are labeled
                    {
                      if (B == C)
                        {
                          L.elem (r, c) = B;
                        }
                      else
                        {
                          int tlabel = std::min (B, C);
                          lset [B] = tlabel;
                          lset [C] = tlabel;
                          L.elem (r, c) = tlabel;
                        }
                    }
                  else if (B) // B is object but C is not
                    {
                      L.elem (r, c) = B;
                    }
                  else if (C) // C is object but B is not
                    {
                      L.elem (r, c) = C;
                    }
                  else // B, C, D not object - new object
                    {
                      // label and put into table
                      ntable ++;
                      L.elem (r, c) = lset [ntable] = ntable;
                    }
                }
              else if (n == 8)
                {
                  // apply 8 connectedness
                  if (B || C || D || E)
                    {
                      int tlabel = B;
                      if (B)
                        {
                          tlabel = B;
                        }
                      else if (C)
                        {
                          tlabel = C;
                        }
                      else if (D)
                        {
                          tlabel = D;
                        }
                      else if (E)
                        {
                          tlabel = E;
                        }

                      L.elem (r, c) = tlabel;

                      if (B && B != tlabel)
                        lset [B] = tlabel;
                      if (C && C != tlabel)
                        lset [C] = tlabel;
                      if (D && D != tlabel)
                        lset [D] = tlabel;
                      if (E && E != tlabel)
                        lset [E] = tlabel;
                    }
                  else
                    {
                      // label and put into table
                      ntable ++;
                      L.elem (r, c) = lset [ntable] = ntable;
                    }
                }
            }
          else
            {
              L.elem (r, c) = NO_OBJECT; // A is not an object so leave it
            }
        }
    }

  // consolidate component table
  for (int i = 0; i <= ntable; i++)
    lset [i] = find (lset, i);

  // run image through the look-up table
  for (int r = 0; r < nr; r++)
    for (int c = 0; c < nc; c++)
      L.elem (r, c) = lset [(int)L.elem (r, c)];

  // count up the objects in the image
  for (int i = 0; i <= ntable; i++)
    lset [i] = 0;

  for (int r = 0; r < nr; r++)
    for (int c = 0; c < nc; c++)
      lset [(int)L.elem (r, c)] ++;

  // number the objects from 1 through n objects
  nobj = 0;
  lset [0] = 0;
  for (int i = 1; i <= ntable; i++)
    if (lset [i] > 0)
      lset [i] = ++nobj;

  // run through the look-up table again
  for (int r = 0; r < nr; r++)
    for (int c = 0; c < nc; c++)
      L.elem (r, c) = lset [(int)L.elem (r, c)];

  rval (0) = L;
  rval (1) = (double)nobj;
  return rval;
}

/*
%!assert(bwlabel(logical([0 1 0; 0 0 0; 1 0 1])),[0 1 0; 0 0 0; 2 0 3]);
*/

File: bwlabel.m
## Copyright (C) 2000  Etienne Grossmann
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

##       [im2,npix,bb] = bwlabel(im,...)
##  
## im  : RxC 0-1 matrix
##
## im2 : RxC int matrix in which the connected regions of im have been
##       numbered. 4-neighborhoods are considered.
##
## npix : 1xQ int number of pixel in each region
##
## bb   : 4xQ int bounding boxes of the regions. Rows are minrow,
##        maxrow, mincol, maxcol.
##
## Since the algorithm is slow, it will report progress if it has to do
## more than 100 loops.
##
## Options :
##
## "verbose"  : Be verbose
## "quiet"    : Don't report anything (overrides "verbose")
## "prudent"  : Check that regions do not touch
## "mins", m  : Minimum size of the regions
## "maxs", m  : Maximum size of the regions
##

## Author:        Etienne Grossmann  <etienne@isr.ist.utl.pt>
## Last modified: January 2000

function [im2,npix,bb] = bwlabel(im,...)

[R,C]=size(im);
tr = 0 ;
				# Loop as little as possible
if R<C, tr = 1; im = im' ; [R,C]=size(im); end

report = C>100 ;

quiet = 0 ;
verbose = 0 ;
prudent = 1 ;
mins = maxs = 0 ;


filename = "bwlabel" ;
opt0 = " verbose quiet " ;
opt1 = " mins maxs " ;
nargin-- ;
read_options ;


if any(im(:)&im(:)!=1),
  printf("bwlabel : im is not binary\n");
  return
end

if quiet, verbose = report = 0 ; end

gsize = 100 ;			# Predicted number of slices

im2 = zeros(R,C);
## keyboard
				# find horizontal up/down going edges
imup =  max(im2, diff([zeros(1,C);im])) ;
imdo = -min(im2, diff([im;zeros(1,C)])) ;

im2 = ones(R,C);
######################################################################
## Find connected regions ############################################

rc = 1 ;			# Counter of region slices 

rnum = ones(1,gsize);		# List of labels of each slice
npix = zeros(1,gsize);		# Sizes of regions (actual size is in
				# region's first slice)
bb = zeros(4,gsize);		# Bounding boxes

if report && !verbose,
  printf("bwlabel : There will be %i loops ... \n%5i ",C,0) ;
end

lrow = zeros(R,1);		# Last treated image column
 
for i=1:C,
  ## i
  t1 = find(imup(:,i)) ;
  t2 = find(imdo(:,i)) ;
  
  nrow = zeros(R,1);		# Next row

  ## if i==42 || i==41,"i==42 || i==41", keyboard; end

  for j = 1:rows(t1(:)) ,	# Loop over slices of i'th column

    rc++ ;			# rc = number of current slice.
    im2(t1(j):t2(j),i) = rc ;
    ## if rc==341, "rc==341",keyboard; end
				# Slices from previous column that touch
				# this slice.
    rr = create_set(lrow(t1(j):t2(j))) ;
    if !isempty(rr) && !rr(1), rr = rr(2:length(rr)) ; end

    if rc>size(rnum,2),		# Get more space (uncertain effect on
				# speed; avoids resizing rnum and npix)
      tmp = 2*(ceil(C*rc/i)-rc+1+rows(t1(:))-j) ;
      sayif(verbose,"bwlabel : (i=%i) Foreseeing %i more regions\n",i,tmp);
      rnum = [ rnum, ones(1,tmp) ] ;
      npix = [ npix,zeros(1,tmp) ] ;
      bb = [ bb,zeros(4,tmp) ] ;
    end

    if isempty(rr),			# New region 
      sayif(verbose,"bwlabel : creating region %i\n",rc);

      r0 = rc ;			# r0 = number of the region

      bb(1,r0) = t1(j) ;
      bb(2,r0) = t2(j) ;
      bb(3:4,r0) = i ;

    else			# Add to already existing region #####

      ## rr
      r0 = rnum(rr(1)) ;

      bb(1,r0) = min(bb(1,r0),t1(j)) ;
      bb(2,r0) = max(bb(2,r0),t2(j)) ;
      bb(4,r0) = i ;

				# Touches region r0
      sayif(verbose,"bwlabel : adding to region %i\n",r0);

				# Touches other regions too
      for k = rr(find(rr!=r0)),	# Loop over other touching regions, that
				# should be merged to the first.
	sayif(verbose,"bwlabel : merging regions %i and %i\n",k,r0);
	rnum(find(rnum==k)) = r0 ;
	npix(r0) = npix(r0) + npix(k) ;
	bb([1,3],r0) = min(bb([1,3],r0)',bb([1,3],k)')' ;
	bb([2,4],r0) = max(bb([2,4],r0)',bb([2,4],k)')' ;
      end
      
    end				# End of add to already existing region
    ## if r0==259 && i==39,"r0==259",keyboard;end
    rnum(rc) = r0 ;

    npix(r0) = npix(r0) + 1+t2(j)-t1(j) ;

    nrow(t1(j):t2(j)) = r0 ;
  end				# End of looping over slices
  lrow = nrow ;
  if report && !verbose,
    printf(".") ;
    if !rem(i,70) && i<C, printf("\n%5i ",i); end
  end
end				# End of looping over columns
## length(regs)
if report && !verbose,
  printf("\n") ;
end
## im2
## [ 1:rc ; rnum(1:rc) ]
keep = ones(1,rc) ; 

if mins, keep = keep & (npix(rnum(1:rc))>=mins) ; end
if maxs, keep = keep & (npix(rnum(1:rc))<=maxs) ; end
keep(1) = 1 ;
keep = find(keep) ;		# Indices of slices to be kept

foo = create_set(rnum(keep)) ;	# Indices of regions to be kept
nr = prod(size(foo)) ;		# Number of regions (including bg)
## lrow = zeros(1,nr) ;
tmp = zeros(1,rc) ;
tmp(foo) = 0:nr-1 ;

bar = zeros(1,rc) ;
bar(keep) = tmp(rnum(keep)) ;

## keyboard

im2 = reshape(bar(im2),R,C) ;

npix = npix(foo(2:nr));
bb = bb(:,foo(2:nr));

if 0,				# Draw bb on the image (will ruin
				# coherence)
  for i=1:nr-1,
    im2([bb(1,i),bb(2,i)] ,bb(3,i):bb(4,i) ) = i ;
    im2( bb(1,i):bb(2,i) ,[bb(3,i),bb(4,i)]) = i ;
  end
end

## keyboard
if prudent,

  sayif(verbose,"bwlabel : Checking coherence\n");

  hcontact = im2 & im2!=[im2(:,2:C),zeros(R,1)] & [im2(:,2:C),zeros(R,1)] ;
  vcontact = im2 & im2!=[im2(2:R,:);zeros(1,C)] & [im2(2:R,:);zeros(1,C)] ;
  ok = 1 ;
  if any(hcontact(:)),
    ok = 0 ;
    printf("bwlabel: Whoa! Found horizontally connected separated regions\n");
  end
  if any(vcontact(:)),
    ok = 0 ;
    printf("bwlabel: Whoa! Found vertically connected separated regions\n");
  end
  if !ok, keyboard ; end
end

## Eventually transpose result
if tr, 
  im2 = im2' ; 
  bb = bb([3,4,1,2],:);
end
